Add type hints to urllib3.response

hramezani opened the pull request:
```
src/urllib3/response.py:152: error: Returning Any from function declared to return "bytes"
src/urllib3/response.py:212: error: Argument 1 to "HTTPHeaderDict" has incompatible type "Optional[Mapping[bytes, bytes]]"; expected "Union[HTTPHeaderDict, Mapping[str, str], Iterable[Tuple[str, str]], HasGettableStringKeys, None]"
src/urllib3/response.py:441: error: Incompatible types in assignment (expression has type "Union[bytes, IO[Any], Iterable[bytes], str]", variable has type "Optional[HTTPResponse]")
src/urllib3/response.py:726: error: Incompatible types in assignment (expression has type "HTTPHeaderDict", variable has type "HTTPMessage")
src/urllib3/response.py:730: error: Argument "headers" to "HTTPResponse" has incompatible type "HTTPMessage"; expected "Optional[Mapping[AnyStr, AnyStr]]"
src/urllib3/response.py:753: error: "Callable[[IOBase], bool]" has no attribute "__get__"
src/urllib3/response.py:753: error: Returning Any from function declared to return "bool"
src/urllib3/response.py:796: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "fp"
src/urllib3/response.py:796: error: Item "None" of "Optional[HTTPResponse]" has no attribute "fp"
src/urllib3/response.py:808: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:808: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:810: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:810: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:813: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:813: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:817: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:817: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:818: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:818: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:822: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:822: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:823: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:823: error: Item "None" of "Optional[HTTPResponse]" has no attribute "_safe_read"
src/urllib3/response.py:825: error: Returning Any from function declared to return "bytes"
src/urllib3/response.py:864: error: Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "fp"
src/urllib3/response.py:864: error: Item "None" of "Optional[HTTPResponse]" has no attribute "fp"
src/urllib3/response.py:86: error: Incompatible types in assignment (expression has type "None", variable has type "bytes")
src/urllib3/response.py:888: error: "HTTPResponse" has no attribute "fp"
src/urllib3/response.py:94: error: Incompatible types in assignment (expression has type "None", variable has type "bytes").```

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -449,19 +469,21 @@ def drain_conn(self) -> None:
 
     @property
-    def data(self):
+    def data(self) -> Optional[Union[bytes, str]]:
```
Should this be `-> bytes`?

hramezani replied:
we have `return self._body` in this function
https://github.com/urllib3/urllib3/blob/main/src/urllib3/response.py#L455

And `_body` can be `Union[str, bytes]` https://github.com/urllib3/urllib3/blob/main/src/urllib3/response.py#L414-L415
Also, The function doesn't have a return(I added a `return None`) 
So, I changed the return type to `Optional[Union[bytes, str]]`

hramezani replied:
after changing the return type to `bytes`, I got:
```
src/urllib3/response.py:475: error: Incompatible return value type (got "Union[str, bytes]", expected "bytes")
src/urllib3/response.py:478: error: Incompatible return value type (got "Optional[bytes]", expected "bytes")
src/urllib3/response.py:480: error: Incompatible return value type (got "None", expected "bytes")
```

We have to ignore them for now.

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -217,39 +237,39 @@ def get_redirect_location(self) -> typing.Union[str, "Literal[False]"]:
 
     @property
-    def data(self) -> bytes:
+    def data(self) -> Optional[Union[bytes, str]]:
```
Should this be -> bytes?

hramezani replied:
to be compatible with the `HTTPResponse.data` property.

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -472,11 +494,11 @@ def tell(self) -> int:
         Set initial length value for Response content if available.
         """
-        length = self.headers.get("content-length")
+        length: Optional[Union[int, str]] = self.headers.get("content-length")
```
Having this typed here as `Optional[Union[int, str]]` instead of `Optional[str]` is causing a `str()` call below to be necessary. Let's have this variable be named `content_length` and keep `length` as it is elsewhere but define it above as `length: Optional[int]`?

hramezani replied:
Then we have `UnboundLocalError: local variable 'length' referenced before assignment` error in 
https://github.com/urllib3/urllib3/blob/2ec06d1185f1642dfe8ff767a9ad3541bd0ddc39/src/urllib3/response.py#L525

when `content-length` is `None`.

sethmlarson replied:
I was going to try to explain in text but it's way faster to just copy-paste and make the edit:

```python
length: Optional[int]
content_length: Optional[str] = self.headers.get("content-length")

if content_length is not None:
    if self.chunked:
        # This Response will fail with an IncompleteRead if it can't be
        # received as chunked. This method falls back to attempt reading
        # the response before raising an exception.
        log.warning(
            "Received response with both Content-Length and "
            "Transfer-Encoding set. This is expressly forbidden "
            "by RFC 7230 sec 3.3.2. Ignoring Content-Length and "
            "attempting to process response as Transfer-Encoding: "
            "chunked."
        )
        return None

    try:
        # RFC 7230 section 3.3.2 specifies multiple content lengths can
        # be sent in a single Content-Length header
        # (e.g. Content-Length: 42, 42). This line ensures the values
        # are all valid ints and that as long as the `set` length is 1,
        # all values are the same. Otherwise, the header is invalid.
        lengths = {int(val) for val in content_length.split(",")}
        if len(lengths) > 1:
            raise InvalidHeader(
                "Content-Length contained multiple "
                "unmatching values (%s)" % content_length
            )
        length = lengths.pop()
    except ValueError:
        length = None
    else:
        if length < 0:
            length = None

else:  # if content_length is None
    length = None

# Convert status to int for comparison
# In some cases, httplib returns a status of "_UNKNOWN"
try:
    status = int(self.status)
except ValueError:
    status = 0

# Check for responses that shouldn't include a body
if status in (204, 304) or 100 <= status < 200 or request_method == "HEAD":
    length = 0

return length
```

Basically the change is the `else: length=None`, does this work?

hramezani replied:
Thanks @sethmlarson, I've replaced your cone block

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -865,7 +896,7 @@ def read_chunked(self, amt=None, decode_content=None):
 
     @property
-    def url(self) -> str:
+    def url(self) -> Optional[Union[str, "Literal[False]"]]:
```
When would this be `False`?

hramezani replied:
changed to `Optional[str]`

sethmlarson commented on src/urllib3/util/retry.py:
```diff
@@ -19,16 +18,21 @@
+    url: Optional[str]
+    error: Optional[Exception]
+    status: Optional[int]
+    redirect_location: Optional[Union[str, "Literal[False]"]]
```
When is this `False`? I think we don't add a `redirect_location` if `False` is given

hramezani replied:
Yes, You are right. I changed it to `Optional[str]` and changed the `increment` function to handle it.

sethmlarson reviewed (changes requested):
Looks pretty good, few comments for you:

hramezani commented on src/urllib3/response.py:
```diff
@@ -173,17 +193,17 @@ class BaseHTTPResponse(io.IOBase):
     REDIRECT_STATUSES = [301, 302, 303, 307, 308]
 
-    DECODER_ERROR_CLASSES = (IOError, zlib.error)
+    DECODER_ERROR_CLASSES: Tuple[Any, ...] = (IOError, zlib.error)
```
Then we have 
```
src/urllib3/response.py:196: error: Incompatible types in assignment (expression has type "Tuple[Type[OSError], Type[error]]", variable has type "Tuple[BaseException, ...]")
src/urllib3/response.py:315: error: Exception type must be derived from BaseException
```

hramezani replied:
Done!

hramezani commented on src/urllib3/exceptions.py:
```diff
@@ -286,7 +295,7 @@ class IncompleteRead(HTTPError, httplib_IncompleteRead):
     partial: int
     expected: int
```
I reverted the constructor to accept `int` and changed the `read` function to fix it.

bluetech reviewed (commented):
Great improvements. I left some suggestions.

sethmlarson commented on src/urllib3/exceptions.py:
```diff
@@ -29,9 +29,13 @@ class HTTPWarning(Warning):
-    def __init__(self, pool: "ConnectionPool", message: str) -> None:
+    def __init__(
+        self,
+        pool: Optional[Union["ConnectionPool", "HTTPConnectionPool"]],
```
The base class has so few attributes, almost to the point of not being useful anymore. Is this why we're doing this construction @hramezani?

hramezani replied:
I think we don't need this change anymore. SO, I've reverted this.

hramezani commented on src/urllib3/response.py:
```diff
@@ -772,14 +802,14 @@ def _update_chunk_length(self):
             self._fp._safe_read(2)  # Toss the CRLF at the end of the chunk.
             self.chunk_left = None
-        elif amt < self.chunk_left:
+        elif self.chunk_left is not None and amt < self.chunk_left:
```
I am not sure but I think it can be `None`. because based on the code [_update_chunk_length](https://github.com/urllib3/urllib3/blob/dfd8b580adb4f87388310c11b5c76d989bd6c4e3/src/urllib3/response.py#L761) will be called before [_handle_chunk](https://github.com/urllib3/urllib3/blob/dfd8b580adb4f87388310c11b5c76d989bd6c4e3/src/urllib3/response.py#L775), and there is:
```
        if self.chunk_left is not None:
            return
```
So, it seems it can be `None`.

hramezani commented on src/urllib3/response.py:
```diff
@@ -173,17 +193,17 @@ class BaseHTTPResponse(io.IOBase):
         self,
         *,
-        headers: typing.Optional[typing.Mapping[typing.AnyStr, typing.AnyStr]] = None,
+        headers: Optional[Mapping[AnyStr, AnyStr]] = None,
```
Then we get:
```
src/urllib3/response.py:211: error: Argument 1 to "HTTPHeaderDict" has incompatible type "Union[Mapping[str, str], Mapping[bytes, bytes], None]"; expected "Union[HTTPHeaderDict, Mapping[str, str], Iterable[Tuple[str, str]], HasGettableStringKeys, None]"
```

hramezani replied:
Done!

hramezani commented on src/urllib3/response.py:
```diff
@@ -217,39 +237,39 @@ def get_redirect_location(self) -> typing.Union[str, "Literal[False]"]:
-        self, amt: int = 2 ** 16, decode_content: typing.Optional[bool] = None
-    ) -> typing.Generator[bytes, None, None]:
+        self, amt: int = 2 ** 16, decode_content: Optional[bool] = None
+    ) -> Generator[bytes, None, None]:
```
Done!

hramezani commented on src/urllib3/response.py:
```diff
@@ -375,23 +395,23 @@ class is also compatible with the Python standard library's :mod:`io`
-        auto_close=True,
-    ):
+        body: HTTPBody = "",
+        headers: Optional[Mapping[AnyStr, AnyStr]] = None,
```
Done!

hramezani commented on src/urllib3/exceptions.py:
```diff
@@ -43,9 +47,14 @@ def __reduce__(self) -> ReduceResult:
-    def __init__(self, pool: "ConnectionPool", url: str, message: str) -> None:
+    def __init__(
+        self,
+        pool: Optional[Union["ConnectionPool", "HTTPConnectionPool"]],
```
This one also reverted.

hramezani commented:
@sethmlarson @bluetech Thanks for your review. I've addressed some of your comments. 
There are some errors like:
```
"HTTPResponse" has no attribute "fp"
Item "HTTPResponse" of "Optional[HTTPResponse]" has no attribute "_safe_read"
```

I think we have to ignore them.

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -221,35 +240,35 @@ def data(self) -> bytes:
-    ) -> bytes:
+        amt: Optional[int] = None,
+        decode_content: Optional[bool] = None,
+    ) -> Generator[bytes, None, None]:
```
Can we do `Iterator[bytes]` here as well?

hramezani replied:
done

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -317,26 +338,24 @@ def readable(self) -> bool:
     def readinto(self, b: bytearray) -> int:
         temp = self.read(len(b))
-        if len(temp) == 0:
+        if temp is None or len(temp) == 0:
```
Since `read()` now returns only `bytes` can we remove the `is None` check here?

hramezani replied:
done

sethmlarson commented on src/urllib3/response.py:
```diff
@@ -852,7 +887,7 @@ def read_chunked(self, amt=None, decode_content=None):
 
             # Chunk content ends with \r\n: discard it.
-            while True:
+            while True and self._fp is not None:
```
We can reduce this to `while self._fp is not None:`?

hramezani replied:
done!

sethmlarson commented on src/urllib3/util/retry.py:
```diff
@@ -461,7 +463,9 @@ def increment(
                 redirect -= 1
             cause = "too many redirects"
-            redirect_location = response.get_redirect_location()
+            response_redirect_location = response.get_redirect_location()
```
Can we remove this if we add a type annotation to `redirect_location` up above like so:

`redirect_location: Optional[str] = None`

hramezani replied:
No, because the return type of `get_redirect_location` is `Union[str, None, Literal[False]]`.
And if we change it to `redirect_location: Union[Optional[str], "Literal[False]"] = None` then we get 
`Argument 5 to "RequestHistory" has incompatible type "Union[str, None, Literal[False]]"; expected "Optional[str]"`
and we need to change `RequestHistory`

sethmlarson reviewed (changes requested):
Few more comments, this is nearly there I think!

sethmlarson commented:
This looks good to me now @hramezani , can you add the type: ignore comments in one commit so I can review them then we can get this merged :)

hramezani commented:
> This looks good to me now @hramezani , can you add the type: ignore comments in one commit so I can review them then we can get this merged :)

Added!

sethmlarson reviewed (approved):
LGTM, thank you!!
